src/plugins/cordova-plugin-device-motion/device-motion-model.js

```diff
@@ -1,47 +1,176 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 
+Math.sign = Math.sign || function (x) {
+    x = +x; // convert to a number
+    if (x === 0 || isNaN(x)) {
+        return Number(x);
+    }
+    return x > 0 ? 1 : -1;
+};
+
 // report interval in milliseconds
 var ACCELEROMETER_REPORT_INTERVAL = 50;
+var G_CONSTANT = 9.81;
+var degToRad = Math.PI / 180;
 
-var _onChangeCallback,
-    _axis = {
-        x: 0,
-        y: 0,
-        z: 0
-    };
+var _onChangeCallback;
+var axisValues = ['x', 'y', 'z'];
+var rotationValues = ['alpha', 'beta', 'gamma'];
+var allValues = axisValues.concat(rotationValues);
+
+var _axis = {
+    x: 0,
+    y: 0,
+    z: 0
+};
+normalizeFromAcceleration(_axis);
 
 function updateAxis(values) {
+    var newValues = applyValues(values, axisValues);
+    normalizeFromAcceleration(newValues);
+    update(newValues);
+}
+
+function updateRotation(values) {
+    var newValues = applyValues(values, rotationValues);
+    normalizeFromRotation(newValues);
+    update(newValues);
+}
+
+function addAxisChangeCallback(callback) {
+    _onChangeCallback = (typeof callback === 'function') ? callback : null;
+}
+
+function applyValues(values, names) {
+    var newValues = JSON.parse(JSON.stringify(_axis));
+    names.forEach(function (name) {
+        if (values.hasOwnProperty(name)) {
+            newValues[name] = values[name];
+        }
+    });
+    return newValues;
+}
+
+function update(values) {
     values = values || {};
 
     var updatedValues = {};
+    var dirty = false;
+
+    allValues.forEach(function (name) {
+        if (values[name] !== _axis[name]) {
+            dirty = true;
+            _axis[name] = values[name];
+            updatedValues[name] = values[name];
+        }
+    });
+
+    // notify if any value has been updated
+    if (dirty && _onChangeCallback) {
+        _onChangeCallback(updatedValues);
+    }
+}
+
+function normalizeFromRotation(values) {
+    values.alpha = constrainAlpha(values.alpha);
+    values.beta = constrainBeta(values.beta);
+    values.gamma = constrainGamma(values.gamma);
+
+    var accel = accelFromRot(values.beta, values.gamma);
+    values.x = accel.x;
+    values.y = accel.y;
+    values.z = accel.z;
+}
+
+function normalizeFromAcceleration(values) {
+    values.x = constrainAcceleration(values.x);
+    values.y = constrainAcceleration(values.y);
+    values.z = constrainAcceleration(values.z);
 
-    if (values.x) {
-        _axis.x = parseFloat(values.x);
-        updatedValues.x = _axis.x;
+    var sinY = -values.y / G_CONSTANT;
```

Comment:
Yeah, it's a valid point. Here was the thought process I went through when I made this change:

My goal was to support keyboard accessibility for this panel as simply as possible. But making the acceleration fields editable opens up a can of worms. First problem was it meant a bunch more work (for the reverse math, and keeping things in sync), and second is the point you raise.

I considered making only the rotation fields editable (and appropriately clamped) - since that is effectively the only functionality available today. But that seemed arbitrarily limiting - there are valid scenarios where you would want to test more extreme acceleration values.

So if we make the acceleration fields editable, we have to deal with the issue you raised, and my conclusion on that was as follows:

In our case, rotation really only exists as a way to define acceleration values (we never report rotation to the app). And going from rotation to acceleration, the magnitude of the acceleration vector will always, of course, be g.

But converting from acceleration to rotation, the magnitude of the acceleration vector could be anything. And if it is anything other than g, there are no possible valid rotation values (assuming a non-accelerating device). If we assume an actual acceleration component plus rotation, how much acceleration and in what direction? Bottom line it seemed to me it didn't really matter... If someone is focusing on acceleration values, the rotation probably isn't that important. And anything we do is wrong one way or another, so we might as well go with what is simplest 😄.

But I'm not strongly tied to that conclusion, so if you feel strongly the other way, let me know.